File: services/auth/src/Auth.Infrastructure/SnowflakeIdGenerator.cs
using Auth.Application.Abstractions;
using Auth.Infrastructure.Options;
using Microsoft.Extensions.Options;

namespace Auth.Infrastructure;

internal sealed class SnowflakeIdGenerator : IIdGenerator
{
    private const int        WorkerIdBits = 10;
    private const int        SequenceBits = 12;
    private const int        TimestampLeftShift = SequenceBits + WorkerIdBits;
    private const long       MaxWorkerId = (1L << WorkerIdBits) - 1;
    private const long       MaxSequence = (1L << SequenceBits) - 1;

    private readonly long   _workerId;
    private readonly long   _epoch;
    private long            _sequence = 0;
    private long            _lastTimestamp = -1;
    private readonly IClock _clock;
    private readonly Lock   _lock = new();

    public SnowflakeIdGenerator(IClock clock, IOptions<SnowFlakeOptions> options)
    {
        var snowflakeOptions = options.Value;
        
        _workerId = snowflakeOptions.WorkerId;
        _epoch = snowflakeOptions.Epoch;
        _clock = clock;

        if (_workerId < 0 || _workerId > MaxWorkerId)
            throw new ArgumentException(
                $"Worker ID must be between 0 and {MaxWorkerId}, but got {_workerId}",
                nameof(snowflakeOptions.WorkerId)
            );
    }

    public long NextId()
    {
        lock (_lock)
        {
            long timestamp = GetCurrentTimestamp();

            if (timestamp < _lastTimestamp)
                throw new InvalidOperationException(
                    "Clock moved backwards. Refusing to generate ID for timestamp " +
                    $"{timestamp} ms. Last known timestamp was {_lastTimestamp} ms."
                );

            if (timestamp == _lastTimestamp)
            {
                _sequence = (_sequence + 1) & MaxSequence;
                if (_sequence == 0) // ? Sequence overflow
                    timestamp = WaitUntilNextMillisecond(_lastTimestamp);
            }
            else
                _sequence = 0;

            _lastTimestamp = timestamp;

            return ((timestamp - _epoch) << TimestampLeftShift)
                | (_workerId << SequenceBits)
                | _sequence;
        }
    }

    private long GetCurrentTimestamp() => _clock.UtcNow.ToUnixTimeMilliseconds();

    private long WaitUntilNextMillisecond(long lastTimestamp)
    {
        long timestamp = GetCurrentTimestamp();
        while (timestamp <= lastTimestamp)
        {
            timestamp = GetCurrentTimestamp();
        }
        return timestamp;
    }
}

File: services/user/src/common/snowflake-id.generator.ts
import { Injectable } from '@nestjs/common';

@Injectable()
export class SnowflakeIdGenerator {
  private static readonly workerIdBits = 10n;
  private static readonly sequenceBits = 12n;
  private static readonly timestampLeftShift =
    SnowflakeIdGenerator.workerIdBits + SnowflakeIdGenerator.sequenceBits;
  private static readonly maxWorkerId =
    (1n << SnowflakeIdGenerator.workerIdBits) - 1n;
  private static readonly maxSequence =
    (1n << SnowflakeIdGenerator.sequenceBits) - 1n;

  private readonly workerId: bigint;
  private readonly epoch: bigint;
  private sequence = 0n;
  private lastTimestamp = -1n;

  constructor() {
    const workerIdRaw = process.env.SNOWFLAKE_WORKER_ID ?? '1';
    const epochRaw = process.env.SNOWFLAKE_EPOCH ?? '1704067200000';
    const workerId = BigInt(workerIdRaw);
    const epoch = BigInt(epochRaw);

    if (workerId < 0n || workerId > SnowflakeIdGenerator.maxWorkerId) {
      throw new Error(
        `SNOWFLAKE_WORKER_ID must be between 0 and ${SnowflakeIdGenerator.maxWorkerId.toString()}`,
      );
    }

    this.workerId = workerId;
    this.epoch = epoch;
  }

  nextId(): string {
    const timestamp = this.currentTimestamp();

    if (timestamp < this.lastTimestamp) {
      throw new Error('Clock moved backwards while generating a snowflake id');
    }

    let nextTimestamp = timestamp;
    if (timestamp === this.lastTimestamp) {
      this.sequence = (this.sequence + 1n) & SnowflakeIdGenerator.maxSequence;
      if (this.sequence === 0n) {
        nextTimestamp = this.waitUntilNextMillisecond(this.lastTimestamp);
      }
    } else {
      this.sequence = 0n;
    }

    this.lastTimestamp = nextTimestamp;

    const id =
      ((nextTimestamp - this.epoch) << SnowflakeIdGenerator.timestampLeftShift) |
      (this.workerId << SnowflakeIdGenerator.sequenceBits) |
      this.sequence;

    return id.toString();
  }

  private currentTimestamp(): bigint {
    return BigInt(Date.now());
  }

  private waitUntilNextMillisecond(lastTimestamp: bigint): bigint {
    let timestamp = this.currentTimestamp();
    while (timestamp <= lastTimestamp) {
      timestamp = this.currentTimestamp();
    }

    return timestamp;
  }
}
